core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala

```diff
@@ -58,63 +62,166 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@transient val
+    (plainExtractors map {case (key, op) => s"Option[$key]" -> returnOption(op)})
+
+  private val extractorsMap: Map[TypeName, Deserializer] = allExtractors withDefaultValue DefaultDeserializer
+
+  val extractors = extractorsMap compose columnTypeNames
+
+  private def opt[X](op: => Any): Option[X] = try {
+    Option(op.asInstanceOf[X])
+  } catch {
+    case ex: Exception => None
+  }
+  // maps data columns to product components
+  val columnMapping: Map[Int, Int] =
+    if (columnTypeNames.size == 1) Map(0->0) else multicolumnMapping
+
+  def multicolumnMapping: Map[Int, Int] = {
+    try {
+      val mappings = for {
+        i <- columnTypeNames.indices
+        name = colNamesInResult(i)
+        j = colNamesInFrame.indexOf(name)
+      } yield (i, j)
+
+      val bads = mappings collect {
+        case (i, j) if j < 0 =>
+          if (i < colNamesInResult.length) colNamesInResult(i) else s"[[$i]] (column of type ${columnTypeNames(i)}"
+      }
+
+      if (bads.nonEmpty) {
+        throw new scala.IllegalArgumentException(s"Missing columns: ${bads mkString ","}")
+      }
+
+      mappings.toMap
+    }
+  }
+
+  class H2ORDDIterator(val keyName: String, val partIndex: Int) extends H2OChunkIterator[A] {
+
+    private def cell(i: Int) = {
+      val j = columnMapping(i)
+      val ex = extractors(i)(converterCtx)
+      val data = ex(j).asInstanceOf[Object]
+//      println(s"@$i/$j = $data")
+      data
+    }
+
+    def extractRow: Option[Array[AnyRef]] = {
+      val rowOpt = opt {
+        val objects: IndexedSeq[Object] = columnTypeNames.indices map cell
+        val row = objects toArray
+
+//        println(s"${converterCtx.rowIdx} -> ${res mkString ":"}")
+        row
+      }
+      converterCtx.increaseRowIdx()
+      rowOpt
+    }
+
+    private var hd: Option[A] = None
+    private var total = 0
 
-      def next(): A = {
-        val data = new Array[Option[Any]](numCols)
-        // FIXME: this is not perfect since ncols does not need to match number of names
-        (0 until numCols).foreach{ idx =>
-          val value = if (converterCtx.isNA(idx)) None
-          else types(idx) match {
-            case q if q == classOf[Integer]           => Some(converterCtx.getInt(idx))
-            case q if q == classOf[java.lang.Long]    => Some(converterCtx.getLong(idx))
-            case q if q == classOf[java.lang.Double]  => Some(converterCtx.getDouble(idx))
-            case q if q == classOf[java.lang.Float]   => Some(converterCtx.getFloat(idx))
-            case q if q == classOf[java.lang.Boolean] => Some(converterCtx.getBoolean(idx))
-            case q if q == classOf[String] => Option(converterCtx.getString(idx))
-            case _ => None
+    override def hasNext = {
+      while (hd.isEmpty && super.hasNext) {
+        hd = readOne()
+        total += 1
+      }
+      hd.isDefined
+    }
+
+    def next(): A = {
+      if (hasNext) {
+        val a = hd.get
```

Comment:
I'm thinking whether these checks are necessary. The iterator is serving us data and basically when  `hd.isDefined` return false, it does not mean that there are no more data in the iterator, but that there was something wrong with this particular row ( If I understand the new logic of H2ORDD right ) => so maybe it would make more sense to throw an exception and write `hasNext` and `next` without the magic of `hd`.

But maybe I'm missing something. What do you think @mmalohlava @mdymczyk ?

I don't like